### PR TITLE
bound the esc key to exit timepicker

### LINKED ISSUE #11332: Escape key doesn't close time picker dialog [5.x]
Selecting the time picker and then hitting escape without changing anything does not close the picker in 5.x. In 4.6 the picker is closed.

### What Grafana version are you using?

5.0.3

### What datasource are you using?

Any

### What OS are you running grafana on?

Linux

### What did you do?

1. Select time picker
2. Press escape

### What was the expected result?

The picker dialog should close

### What happened instead?

It did not close

### PR DESCRIPTION
**Pull request by Ijin08**
fixes #11332

### CODE REVIEW

**Review by marefr — CHANGES_REQUESTED** [review]
Basically, it works to close the time picker with escape when opened. Have no feedback on code, looks good.

If I'm in view/edit mode of a panel, open timepicker and then escape I would expect the timepicker to be closed and still have view/edit mode, but right now it closes timepicker and edit/view mode - **needs to be fixed**.

However I agree on the comment in #11332 :
> in 4x you could close the picker by clicking anywhere not in the popup. In 5x it stays open until you hit either 'apply' or the range button again.

Would be **nice to fix this now as well (if possible)**, if not we should create a new issue for that.

**Comment by Ijin08** [comment]
Thanks for feedback!

About the comment there is an issue already: #10757

**Comment by marefr** [comment]
Okay. Would like that issue to be part of some milestone at least. Can discuss it with team...

**Review by marefr — CHANGES_REQUESTED** [review]
Earlier problem with going into edit/view mode is fixed :+1: 

Found another issue. Included animated gif below but you can't see when I press the esc key. However, if you open the timepicker and click on the time it will close which is good. However if I now go in to edit/view mode and press escape the time picker shows up. Think you need to publish an event from timepicker when it's closing so that you can sync the state (closed/open) in keybindingSrv.ts.

image

File: public/app/features/dashboard/timepicker/timepicker.ts [hunk 1]
```diff
@@ -96,13 +96,12 @@ export class TimePickerCtrl {
   }
 
   openDropdown() {
+    this.$rootScope.appEvent('escTimepicker');
```
Comment (marefr) [inline public/app/features/dashboard/timepicker/timepicker.ts]: This looks weird. The timepicker should not care about if esc or any other key is pressed. We want the timepicker to be as dumb as possible and only care about it's internal functionality.

File: public/app/features/dashboard/timepicker/timepicker.ts [hunk 2]
```diff
@@ -96,13 +96,12 @@ export class TimePickerCtrl {
...
     if (this.isOpen) {
       this.isOpen = false;
```
Comment (marefr) [inline public/app/features/dashboard/timepicker/timepicker.ts]: Here we probably want to emit appEvent that timepicker is closed, i.e. `timepicker-closed`. I think a function closeDropdown would be nice where bot `this.isOpen = false;` and the appEvent is emited. Then you can call this function at the end of `setRelativeFilter` and `applyCustom`.

File: public/app/features/dashboard/timepicker/timepicker.ts [hunk 3]
```diff
@@ -118,6 +117,7 @@ export class TimePickerCtrl {
   }
 
   applyCustom() {
+    this.$rootScope.appEvent('escTimepicker');
```
Comment (marefr) [inline public/app/features/dashboard/timepicker/timepicker.ts]: See comment above about timepicker-closed

File: public/app/features/dashboard/timepicker/timepicker.ts [hunk 4]
```diff
@@ -139,6 +139,7 @@ export class TimePickerCtrl {
   }
 
   setRelativeFilter(timespan) {
+    this.$rootScope.appEvent('escTimepicker');
```
Comment (marefr) [inline public/app/features/dashboard/timepicker/timepicker.ts]: See comment above about `timepicker-closed`

File: public/app/features/dashboard/timepicker/timepicker.ts [hunk 5]
```diff
@@ -118,6 +117,7 @@ export class TimePickerCtrl {
   }
```
Comment (marefr) [inline public/app/features/dashboard/timepicker/timepicker.ts]: Here we probably want to emit appEvent that timepicker is opened, i.e. `timepicker-opened`.

File: public/app/core/services/keybindingSrv.ts [hunk 6]
```diff
@@ -23,7 +23,13 @@ export class KeybindingSrv {
 
     this.setupGlobal();
     appEvents.on('show-modal', () => (this.modalOpen = true));
-    $rootScope.onAppEvent('openTimepicker', () => (this.timepickerOpen = true));
+    $rootScope.onAppEvent('escTimepicker', () => {
```
Comment (marefr) [inline public/app/core/services/keybindingSrv.ts]: Handle `timepicker-closed` and `timepicker-opened` events here instead

**Review by marefr — CHANGES_REQUESTED** [review]
It works as expected, but I would like some changes of the code.

**Review by marefr — APPROVED** [review]

**Pull request merged**